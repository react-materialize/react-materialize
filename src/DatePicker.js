import React, { useMemo, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import TextInput from './TextInput';
import idgen from './idgen';

const DatePicker = props => {
  const id = useMemo(() => props.id || `datepicker${idgen()}`, [props.id]);

  useEffect(() => {
    const { onChange } = props;
    const options = onChange
      ? { ...props.options, onSelect: onChange }
      : props.options;
    const elem = document.getElementById(id);
    const instance = M.Datepicker.init(elem, options);

    return () => {
      if (instance) {
        instance.destroy();
      }
    };
  }, [props]);

  return <TextInput id={id} inputClassName="datepicker" {...props} />;
};

DatePicker.propTypes = {
  /**
   * Event called when Time has been selected
   */
  onChange: PropTypes.func,
  /**
   * id passed to datepicker, also used for init method
   */
  id: PropTypes.string,
  /**
   * options passed to init method
   * more info: https://materializecss.com/pickers.html#date-picker
   */
  options: PropTypes.shape({
    /**
     * Automatically close picker when date is selected.
     */
    autoClose: PropTypes.bool,
    /**
     * The date output format for the input field value.
     * @default 'mmm dd, yyyy'
     */
    format: PropTypes.string,
    /**
     * Used to create date object from current input string.
     * @default null
     */
    parse: PropTypes.func,
    /**
     * The initial date to view when first opened.
     * @default null
     */
    defaultDate: PropTypes.any,
    /**
     * Make the defaultDate the initial selected value.
     * @default false
     */
    setDefaultDate: PropTypes.bool,
    /**
     * Prevent selection of any date on the weekend.
     * @default false
     */
    disableWeekends: PropTypes.bool,
    /**
     * Custom function to disable certain days.
     * @default null
     */
    disableDayFn: PropTypes.func,
    /**
     * First day of week (0: Sunday, 1: Monday etc).
     * @default 0
     */
    firstDay: PropTypes.number,
    /**
     * The earliest date that can be selected.
     * @default null
     */
    minDate: PropTypes.any,
    /**
     * The latest date that can be selected.
     * @default null
     */
    maxDate: PropTypes.any,
    /**
     * Number of years either side, or array of upper/lower range.
     * @default 10
     */
    yearRange: PropTypes.oneOfType([PropTypes.number, PropTypes.array]),
    /**
     * Changes Datepicker to RTL.
     * @default false
     */
    isRTL: PropTypes.bool,
    /**
     * Show month after year in Datepicker title.
     * @default false
     */
    showMonthAfterYear: PropTypes.bool,
    /**
     * Render days of the calendar grid that fall in the next or previous month.
     * @default false
     */
    showDaysInNextAndPreviousMonths: PropTypes.bool,
    /**
     * Specify a DOM element to render the calendar in, by default it will be placed before the input.
     * @default null
     */
    container: PropTypes.any,
    /**
     * Show the clear button in the datepicker.
     * @default false
     */
    showClearBtn: PropTypes.bool,
    /**
     * Internationalization options.
     * @default See i18n documentation.
     */
    i18n: PropTypes.shape({
      cancel: PropTypes.string,
      clear: PropTypes.string,
      done: PropTypes.string,
      previousMonth: PropTypes.string,
      nextMonth: PropTypes.string,
      months: PropTypes.arrayOf(PropTypes.string),
      monthsShort: PropTypes.arrayOf(PropTypes.string),
      weekdays: PropTypes.arrayOf(PropTypes.string),
      weekdaysShort: PropTypes.arrayOf(PropTypes.string),
      weekdaysAbbrev: PropTypes.arrayOf(PropTypes.string)
    }),
    /**
     * An array of string returned by `Date.toDateString()`, indicating there are events in the specified days.
     * @default []
     */
    events: PropTypes.arrayOf(PropTypes.string),
    /**
     * Callback function when date is selected, first parameter is the newly selected date.
     * @default null
     */
    onSelect: PropTypes.func,
    /**
     * Callback function when Datepicker is opened.
     * @default null
     */
    onOpen: PropTypes.func,
    /**
     * Callback function when Datepicker is closed.
     * @default null
     */
    onClose: PropTypes.func,
    /**
     * Callback function when Datepicker HTML is refreshed.
     * @default null
     */
    onDraw: PropTypes.func
  })
};

DatePicker.defaultProps = {
  options: {
    autoClose: false,
    format: 'mmm dd, yyyy',
    parse: null,
    defaultDate: null,
    setDefaultDate: false,
    disableWeekends: false,
    disableDayFn: null,
    firstDay: 0,
    minDate: null,
    maxDate: null,
    yearRange: 10,
    isRTL: false,
    showMonthAfterYear: false,
    showDaysInNextAndPreviousMonths: false,
    container: null,
    showClearBtn: false,
    i18n: {
      cancel: 'Cancel',
      clear: 'Clear',
      done: 'Ok',
      previousMonth: '‹',
      nextMonth: '›',
      months: [
        'January',
        'February',
        'March',
        'April',
        'May',
        'June',
        'July',
        'August',
        'September',
        'October',
        'November',
        'December'
      ],
      monthsShort: [
        'Jan',
        'Feb',
        'Mar',
        'Apr',
        'May',
        'Jun',
        'Jul',
        'Aug',
        'Sep',
        'Oct',
        'Nov',
        'Dec'
      ],
      weekdays: [
        'Sunday',
        'Monday',
        'Tuesday',
        'Wednesday',
        'Thursday',
        'Friday',
        'Saturday'
      ],
      weekdaysShort: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
      weekdaysAbbrev: ['S', 'M', 'T', 'W', 'T', 'F', 'S']
    },
    events: [],
    onSelect: null,
    onOpen: null,
    onClose: null,
    onDraw: null
  }
};

export default DatePicker;
