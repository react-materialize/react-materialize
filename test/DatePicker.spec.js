import React from 'react';
import { shallow, mount } from 'enzyme';
import DatePicker from '../src/DatePicker';
import mocker from './helper/new-mocker';

describe('<DatePicker />', () => {
  let wrapper;

  test('renders', () => {
    wrapper = shallow(<DatePicker />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders with a label', () => {
    wrapper = shallow(<DatePicker label="DatePicker label" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('accepts external `id` value', () => {
    wrapper = shallow(<DatePicker id="ID" />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('initialises', () => {
    const datePickerInitMock = jest.fn();
    const datePickerInstanceDestroyMock = jest.fn();
    const datePickerMock = {
      init: (el, options) => {
        datePickerInitMock(options);
        return {
          destroy: datePickerInstanceDestroyMock
        };
      }
    };

    const restore = mocker('Datepicker', datePickerMock);

    beforeEach(() => {
      datePickerInitMock.mockClear();
      datePickerInstanceDestroyMock.mockClear();
    });

    afterAll(() => {
      restore();
    });

    test('uses default options if none are given', () => {
      wrapper = mount(<DatePicker />);

      expect(datePickerInitMock).toHaveBeenCalledWith({
        autoClose: false,
        format: 'mmm dd, yyyy',
        parse: null,
        defaultDate: null,
        setDefaultDate: false,
        disableWeekends: false,
        disableDayFn: null,
        firstDay: 0,
        minDate: null,
        maxDate: null,
        yearRange: 10,
        isRTL: false,
        showMonthAfterYear: false,
        showDaysInNextAndPreviousMonths: false,
        container: null,
        showClearBtn: false,
        i18n: {
          cancel: 'Cancel',
          clear: 'Clear',
          done: 'Ok',
          previousMonth: '‹',
          nextMonth: '›',
          months: [
            'January',
            'February',
            'March',
            'April',
            'May',
            'June',
            'July',
            'August',
            'September',
            'October',
            'November',
            'December'
          ],
          monthsShort: [
            'Jan',
            'Feb',
            'Mar',
            'Apr',
            'May',
            'Jun',
            'Jul',
            'Aug',
            'Sep',
            'Oct',
            'Nov',
            'Dec'
          ],
          weekdays: [
            'Sunday',
            'Monday',
            'Tuesday',
            'Wednesday',
            'Thursday',
            'Friday',
            'Saturday'
          ],
          weekdaysShort: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
          weekdaysAbbrev: ['S', 'M', 'T', 'W', 'T', 'F', 'S']
        },
        events: [],
        onSelect: null,
        onOpen: null,
        onClose: null,
        onDraw: null
      });
    });
  });
});
