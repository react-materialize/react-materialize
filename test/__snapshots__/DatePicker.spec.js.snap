// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> accepts external \`id\` value 1`] = `
<TextInput
  id="ID"
  inputClassName="datepicker"
  options={
    Object {
      "autoClose": false,
      "container": null,
      "defaultDate": null,
      "disableDayFn": null,
      "disableWeekends": false,
      "events": Array [],
      "firstDay": 0,
      "format": "mmm dd, yyyy",
      "i18n": Object {
        "cancel": "Cancel",
        "clear": "Clear",
        "done": "Ok",
        "months": Array [
          "January",
          "February",
          "March",
          "April",
          "May",
          "June",
          "July",
          "August",
          "September",
          "October",
          "November",
          "December",
        ],
        "monthsShort": Array [
          "Jan",
          "Feb",
          "Mar",
          "Apr",
          "May",
          "Jun",
          "Jul",
          "Aug",
          "Sep",
          "Oct",
          "Nov",
          "Dec",
        ],
        "nextMonth": "›",
        "previousMonth": "‹",
        "weekdays": Array [
          "Sunday",
          "Monday",
          "Tuesday",
          "Wednesday",
          "Thursday",
          "Friday",
          "Saturday",
        ],
        "weekdaysAbbrev": Array [
          "S",
          "M",
          "T",
          "W",
          "T",
          "F",
          "S",
        ],
        "weekdaysShort": Array [
          "Sun",
          "Mon",
          "Tue",
          "Wed",
          "Thu",
          "Fri",
          "Sat",
        ],
      },
      "isRTL": false,
      "maxDate": null,
      "minDate": null,
      "onClose": null,
      "onDraw": null,
      "onOpen": null,
      "onSelect": null,
      "parse": null,
      "setDefaultDate": false,
      "showClearBtn": false,
      "showDaysInNextAndPreviousMonths": false,
      "showMonthAfterYear": false,
      "yearRange": 10,
    }
  }
/>
`;

exports[`<DatePicker /> renders 1`] = `
<TextInput
  id="datepickermockId"
  inputClassName="datepicker"
  options={
    Object {
      "autoClose": false,
      "container": null,
      "defaultDate": null,
      "disableDayFn": null,
      "disableWeekends": false,
      "events": Array [],
      "firstDay": 0,
      "format": "mmm dd, yyyy",
      "i18n": Object {
        "cancel": "Cancel",
        "clear": "Clear",
        "done": "Ok",
        "months": Array [
          "January",
          "February",
          "March",
          "April",
          "May",
          "June",
          "July",
          "August",
          "September",
          "October",
          "November",
          "December",
        ],
        "monthsShort": Array [
          "Jan",
          "Feb",
          "Mar",
          "Apr",
          "May",
          "Jun",
          "Jul",
          "Aug",
          "Sep",
          "Oct",
          "Nov",
          "Dec",
        ],
        "nextMonth": "›",
        "previousMonth": "‹",
        "weekdays": Array [
          "Sunday",
          "Monday",
          "Tuesday",
          "Wednesday",
          "Thursday",
          "Friday",
          "Saturday",
        ],
        "weekdaysAbbrev": Array [
          "S",
          "M",
          "T",
          "W",
          "T",
          "F",
          "S",
        ],
        "weekdaysShort": Array [
          "Sun",
          "Mon",
          "Tue",
          "Wed",
          "Thu",
          "Fri",
          "Sat",
        ],
      },
      "isRTL": false,
      "maxDate": null,
      "minDate": null,
      "onClose": null,
      "onDraw": null,
      "onOpen": null,
      "onSelect": null,
      "parse": null,
      "setDefaultDate": false,
      "showClearBtn": false,
      "showDaysInNextAndPreviousMonths": false,
      "showMonthAfterYear": false,
      "yearRange": 10,
    }
  }
/>
`;

exports[`<DatePicker /> renders with a label 1`] = `
<TextInput
  id="datepickermockId"
  inputClassName="datepicker"
  label="DatePicker label"
  options={
    Object {
      "autoClose": false,
      "container": null,
      "defaultDate": null,
      "disableDayFn": null,
      "disableWeekends": false,
      "events": Array [],
      "firstDay": 0,
      "format": "mmm dd, yyyy",
      "i18n": Object {
        "cancel": "Cancel",
        "clear": "Clear",
        "done": "Ok",
        "months": Array [
          "January",
          "February",
          "March",
          "April",
          "May",
          "June",
          "July",
          "August",
          "September",
          "October",
          "November",
          "December",
        ],
        "monthsShort": Array [
          "Jan",
          "Feb",
          "Mar",
          "Apr",
          "May",
          "Jun",
          "Jul",
          "Aug",
          "Sep",
          "Oct",
          "Nov",
          "Dec",
        ],
        "nextMonth": "›",
        "previousMonth": "‹",
        "weekdays": Array [
          "Sunday",
          "Monday",
          "Tuesday",
          "Wednesday",
          "Thursday",
          "Friday",
          "Saturday",
        ],
        "weekdaysAbbrev": Array [
          "S",
          "M",
          "T",
          "W",
          "T",
          "F",
          "S",
        ],
        "weekdaysShort": Array [
          "Sun",
          "Mon",
          "Tue",
          "Wed",
          "Thu",
          "Fri",
          "Sat",
        ],
      },
      "isRTL": false,
      "maxDate": null,
      "minDate": null,
      "onClose": null,
      "onDraw": null,
      "onOpen": null,
      "onSelect": null,
      "parse": null,
      "setDefaultDate": false,
      "showClearBtn": false,
      "showDaysInNextAndPreviousMonths": false,
      "showMonthAfterYear": false,
      "yearRange": 10,
    }
  }
/>
`;
